File: stackunderflow/adapters/qwen.py
"""Qwen Code session adapter.

Reads Qwen CLI session JSONL files at::

    $QWEN_DATA_DIR/projects/{project}/chats/*.jsonl
    ~/.qwen/projects/{project}/chats/*.jsonl  (default)

Each entry is one JSON object on its own line carrying the shape::

    {
      uuid, sessionId, timestamp,
      type: 'user' | 'assistant',
      model?,
      message: {
        role,
        parts: [{ text, thought?, functionCall: { name, args } }]
      },
      usageMetadata: {
        promptTokenCount, candidatesTokenCount,
        thoughtsTokenCount, cachedContentTokenCount
      }
    }

One ``Record`` is yielded per assistant entry that carries
``usageMetadata``; user entries also produce records (with zero tokens)
for conversation accounting. ``seq`` is the byte offset of the line
start — same convention as the Codex / Claude JSONL adapters, so
``read(ref, since_offset=N)`` resumes by ``fh.seek(N)`` and the
storage-aware contract test holds.

Token normalization (canonical 4-key shape):

* ``input_tokens`` = ``promptTokenCount - cachedContentTokenCount``
  (cached counts inside ``promptTokenCount``; canonical input is fresh
  input only — same convention as OpenAI / Anthropic).
* ``output_tokens`` = ``candidatesTokenCount + thoughtsTokenCount``
  (reasoning ("thoughts") rolls into the billable output column —
  matches the convention used for OpenAI reasoning tokens).
* ``cache_read_tokens`` = ``cachedContentTokenCount``.
* ``cache_create_tokens`` = ``0`` (Qwen does not surface cache writes).

macOS-only path constants in v1; Windows / Linux are documented at the
constant definitions but ``# untested`` per spec §5.

Defensive sizing: JSONL chats larger than ``MAX_SESSION_FILE_BYTES``
(128 MB; see ``stackunderflow/adapters/_streaming.py``) are **skipped
with a logged warning** rather than parsed. Smaller files stream
line-by-line.
"""

from __future__ import annotations

import json
import logging
import os
from collections.abc import Iterator
from pathlib import Path

from ._streaming import iter_jsonl_lines
from .base import Record, SessionRef

_log = logging.getLogger(__name__)

# Files bigger than this trigger a warning but are still parsed. Same
# threshold as the Codex adapter — keeps surprise-large files visible
# in the logs without aborting the read.
_LARGE_FILE_BYTES = 64 * 1024 * 1024

# Qwen tool name -> canonical cross-source tool label. Mirror the small
# set Codex maps; unknown names pass through untouched.
_TOOL_NAME_MAP = {
    "shell": "Bash",
    "execute": "Bash",
    "exec_command": "Bash",
    "run_command": "Bash",
    "read_file": "Read",
    "edit_file": "Edit",
    "write_file": "Edit",
    "apply_diff": "Edit",
    "list_directory": "Glob",
    "glob": "Glob",
    "grep": "Grep",
    "search": "Grep",
}

_DEFAULT_MODEL = "qwen-auto"


def _qwen_root() -> Path:
    """Resolve the Qwen projects root, honouring ``$QWEN_DATA_DIR``.

    Codeburn `qwen-parser.ts` checks ``QWEN_DATA_DIR`` first; that
    behaviour matters for users with sandboxed installs. We mirror it
    exactly so a single test can swap the env var instead of needing a
    constructor override.
    """
    env = os.environ.get("QWEN_DATA_DIR")
    if env:
        return Path(env) / "projects"
    return Path.home() / ".qwen" / "projects"


class QwenAdapter:
    """Source adapter for the Qwen Code CLI."""

    name = "qwen"

    def __init__(self, projects_root: Path | None = None) -> None:
        # ``projects_root`` is overridable so tests can point at synthetic
        # fixtures without monkey-patching ``Path.home()`` or the env.
        self._root = projects_root if projects_root is not None else _qwen_root()

    # ── enumeration ───────────────────────────────────────────────────

    def enumerate(self) -> Iterator[SessionRef]:
        root = self._root
        if not root.is_dir():
            # Not installed / never used — clean no-op rather than raise.
            return

        for project_dir in sorted(p for p in root.iterdir() if p.is_dir()):
            chats_dir = project_dir / "chats"
            if not chats_dir.is_dir():
                continue
            for fp in sorted(chats_dir.glob("*.jsonl")):
                try:
                    stat = fp.stat()
                except OSError as exc:
                    _log.warning("Cannot stat Qwen chat %s: %s", fp, exc)
                    continue

                if stat.st_size > _LARGE_FILE_BYTES:
                    _log.warning(
                        "Qwen chat %s is %d bytes (>%d); reading anyway",
                        fp, stat.st_size, _LARGE_FILE_BYTES,
                    )

                yield SessionRef(
                    provider=self.name,
                    project_slug=project_dir.name,
                    # ``session_id`` is finalised in ``read()`` from the
                    # first entry's ``sessionId`` field — we use the
                    # filename stem here so the registry/store has a
                    # deterministic id even before the file is opened.
                    session_id=fp.stem,
                    file_path=fp,
                    file_mtime=stat.st_mtime,
                    file_size=stat.st_size,
                    source_kind="file",
                    source_hint=None,
                )

    # ── reading ───────────────────────────────────────────────────────

    def read(self, ref: SessionRef, *, since_offset: int = 0) -> Iterator[Record]:
        # ``iter_jsonl_lines`` enforces the 128 MB defensive cap and
        # streams line-by-line; oversized chats are skipped with a
        # warning rather than parsed.
        for line_offset, raw_line in iter_jsonl_lines(
            ref.file_path, since_offset=since_offset,
        ):
            # ``since_offset == 0`` means "fresh read, yield
            # everything". Otherwise the caller already saw the
            # record at exactly ``since_offset`` so skip it.
            if since_offset > 0 and line_offset <= since_offset:
                continue
            stripped = raw_line.strip()
            if not stripped:
                continue
            try:
                entry = json.loads(stripped)
            except (json.JSONDecodeError, ValueError) as exc:
                _log.debug(
                    "Skipping malformed Qwen JSON line in %s: %s",
                    ref.file_path, exc,
                )
                continue
            if not isinstance(entry, dict):
                continue

            record = self._record_from_entry(entry, ref=ref, seq=line_offset)
            if record is not None:
                yield record

    # ── internals ─────────────────────────────────────────────────────

    def _record_from_entry(
        self,
        entry: dict,
        *,
        ref: SessionRef,
        seq: int,
    ) -> Record | None:
        etype = entry.get("type")
        if etype not in ("user", "assistant"):
            return None
        role = etype  # 'user' / 'assistant'

        message = entry.get("message") if isinstance(entry.get("message"), dict) else {}
        parts = message.get("parts") if isinstance(message, dict) else None
        text = _text_from_parts(parts)
        tools = _tools_from_parts(parts)

        usage = entry.get("usageMetadata")
        tokens = _normalize_usage(usage)

        session_id = str(entry.get("sessionId") or ref.session_id)
        raw_model = entry.get("model")
        if not isinstance(raw_model, str) or not raw_model:
            # A non-string model (dict / list / number) would poison the
            # Record contract and crash the store write downstream.
            raw_model = None
        model = raw_model or (_DEFAULT_MODEL if role == "assistant" else None)
        timestamp = str(entry.get("timestamp") or "")
        uuid = str(entry.get("uuid") or f"{session_id}:{seq}")

        return Record(
            provider=self.name,
            session_id=session_id,
            seq=seq,
            timestamp=timestamp,
            role=role,
            model=model,
            input_tokens=tokens["input"],
            output_tokens=tokens["output"],
            cache_create_tokens=tokens["cache_creation"],
            cache_read_tokens=tokens["cache_read"],
            content_text=text,
            tools=tools,
            cwd=None,
            is_sidechain=False,
            uuid=uuid,
            parent_uuid=None,
            raw=entry,
        )


# ── helpers ───────────────────────────────────────────────────────────


def _text_from_parts(parts: object) -> str:
    """Concatenate every ``.text`` field across content parts.

    Parts marked ``thought=True`` are reasoning traces — we still
    include their text in ``content_text`` for downstream search /
    classification, mirroring how the Claude adapter keeps
    ``thinking`` blocks visible.
    """
    if not isinstance(parts, list):
        return ""
    pieces: list[str] = []
    for part in parts:
        if isinstance(part, dict):
            text = part.get("text")
            if isinstance(text, str) and text:
                pieces.append(text)
        elif isinstance(part, str):
            pieces.append(part)
    return "\n".join(pieces)


def _tools_from_parts(parts: object) -> tuple[str, ...]:
    """Extract tool names from ``functionCall`` blocks in parts.

    Unknown names pass through unmapped so new Qwen tools remain visible
    in the dashboard — same convention as the Codex adapter.
    """
    if not isinstance(parts, list):
        return ()
    names: list[str] = []
    for part in parts:
        if not isinstance(part, dict):
            continue
        fc = part.get("functionCall")
        if not isinstance(fc, dict):
            continue
        name = fc.get("name")
        if not isinstance(name, str) or not name:
            continue
        names.append(_TOOL_NAME_MAP.get(name, name))
    return tuple(names)


def _normalize_usage(usage: object) -> dict[str, int]:
    """Flatten Qwen ``usageMetadata`` into the canonical 4-key shape.

    See the module docstring for the full rule. Missing / malformed
    fields default to 0 so a partial entry still produces a valid
    Record.
    """
    if not isinstance(usage, dict):
        return {"input": 0, "output": 0, "cache_creation": 0, "cache_read": 0}

    prompt = _safe_int(usage.get("promptTokenCount"))
    candidates = _safe_int(usage.get("candidatesTokenCount"))
    thoughts = _safe_int(usage.get("thoughtsTokenCount"))
    cached = _safe_int(usage.get("cachedContentTokenCount"))

    return {
        "input": max(prompt - cached, 0),
        "output": candidates + thoughts,
        "cache_creation": 0,
        "cache_read": cached,
    }


def _safe_int(value: object) -> int:
    """Coerce ``value`` to a non-negative int, defaulting to 0 on failure."""
    if value is None or value == "":
        return 0
    try:
        out = int(value)
    except (TypeError, ValueError, OverflowError):
        # OverflowError: JSON like ``1e999`` parses to float('inf').
        return 0
    return max(out, 0)

File: rust/crates/stax-adapters/src/qwen.rs
//! Qwen Code — the port of `python-legacy: adapters/qwen.py`.
//!
//! Chats live at `$QWEN_DATA_DIR/projects/<project>/chats/*.jsonl`, defaulting
//! to `~/.qwen/projects`. One JSON object per line; `seq` is the byte offset of
//! the line start, so this is the plain byte-offset JSONL shape the Claude and
//! Codex adapters already established — a resumed read is a `seek`.
//!
//! ## The token flattening
//!
//! Qwen writes Google's `usageMetadata` shape, where cached input is counted
//! *inside* `promptTokenCount` and reasoning ("thoughts") is billed separately.
//! [`normalize_usage`] flattens that into the canonical four slots the store
//! keeps for every provider: fresh input only, reasoning folded into output,
//! cached input under `cache_read`, and `cache_creation` at 0 because Qwen never
//! surfaces a cache write. Identical in shape to
//! [`crate::codex::canonicalize_openai_usage`] — same convention, different key
//! names.
//!
//! ## Environment, injected
//!
//! `$QWEN_DATA_DIR` relocates the whole tree. Python resolves it once in
//! `__init__`, so [`QwenAdapter::new`] does too; [`QwenAdapter::with_projects_root`]
//! injects a root instead, which is how the tests and the parity harness avoid
//! `set_var` (forbidden: Rust 2024 makes it `unsafe`, the workspace forbids
//! `unsafe`).

use std::ffi::OsStr;
use std::path::{Path, PathBuf};

use serde_json::{Map, Value};

use crate::base::{
    Record, SessionRef, SourceAdapter, Speed, file_name, file_stem, glob_suffix, home_dir,
    read_dir_sorted, stat_ref_fields,
};
use crate::jsonl::{JsonlLines, parse_json, py_bytes_strip};
use crate::pyval;

/// The provider key.
pub const NAME: &str = "qwen";

/// The `$QWEN_DATA_DIR` override (`qwen.py:97`).
pub const DATA_DIR_ENV: &str = "QWEN_DATA_DIR";

/// Model stamped on an assistant turn that records none (`_DEFAULT_MODEL`).
pub const DEFAULT_MODEL: &str = "qwen-auto";

/// Chats bigger than this are warned about but still parsed
/// (`_LARGE_FILE_BYTES`). The hard 128 MB skip lives in
/// [`crate::jsonl::MAX_SESSION_FILE_BYTES`].
pub const LARGE_FILE_BYTES: u64 = 64 * 1024 * 1024;

/// Qwen tool name → canonical cross-source tool label (`_TOOL_NAME_MAP`).
///
/// Unknown names pass through untouched so a new Qwen tool stays visible until
/// it is classified.
pub const TOOL_NAME_MAP: [(&str, &str); 12] = [
    ("shell", "Bash"),
    ("execute", "Bash"),
    ("exec_command", "Bash"),
    ("run_command", "Bash"),
    ("read_file", "Read"),
    ("edit_file", "Edit"),
    ("write_file", "Edit"),
    ("apply_diff", "Edit"),
    ("list_directory", "Glob"),
    ("glob", "Glob"),
    ("grep", "Grep"),
    ("search", "Grep"),
];

/// The Qwen projects root, with the environment injected (`_qwen_root`).
///
/// `$QWEN_DATA_DIR/projects` when the variable is set and non-empty (Python's
/// `if env:` is a plain truthiness test — it does *not* strip, unlike
/// `CLAUDE_CONFIG_DIR`), else `<home>/.qwen/projects`.
#[must_use]
pub fn resolve_projects_root(data_dir: Option<&OsStr>, home: Option<&Path>) -> PathBuf {
    match data_dir.filter(|value| !value.is_empty()) {
        Some(value) => Path::new(value).join("projects"),
        None => home.map_or_else(
            || PathBuf::from(".qwen").join("projects"),
            |home| home.join(".qwen").join("projects"),
        ),
    }
}

/// The Qwen source adapter (`QwenAdapter`).
#[derive(Debug, Clone)]
pub struct QwenAdapter {
    root: PathBuf,
}

impl Default for QwenAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl QwenAdapter {
    /// Read the live environment, as Python's `__init__` does — once, at
    /// construction.
    #[must_use]
    pub fn new() -> Self {
        Self {
            root: resolve_projects_root(
                std::env::var_os(DATA_DIR_ENV).as_deref(),
                home_dir().as_deref(),
            ),
        }
    }

    /// Inject the projects root — the constructor parameter Python already has.
    #[must_use]
    pub fn with_projects_root(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    /// The projects root this adapter reads.
    #[must_use]
    pub fn projects_root(&self) -> &Path {
        &self.root
    }
}

impl SourceAdapter for QwenAdapter {
    fn name(&self) -> &str {
        NAME
    }

    fn enumerate(&self) -> Vec<SessionRef> {
        if !self.root.is_dir() {
            // Not installed / never used — clean no-op rather than raise.
            return Vec::new();
        }
        let mut out = Vec::new();
        for project_dir in read_dir_sorted(&self.root) {
            if !project_dir.is_dir() {
                continue;
            }
            let chats_dir = project_dir.join("chats");
            if !chats_dir.is_dir() {
                continue;
            }
            for path in glob_suffix(&chats_dir, ".jsonl") {
                // Python warns and continues on OSError here; `None` is the
                // same outcome without the crash risk.
                let Some((mtime, size)) = stat_ref_fields(&path) else {
                    continue;
                };
                if size > LARGE_FILE_BYTES {
                    // LOG: python warns "Qwen chat %s is %d bytes; reading anyway".
                }
                out.push(SessionRef::file(
                    NAME,
                    file_name(&project_dir),
                    // Finalised in `read()` from the entry's `sessionId`; the
                    // stem keeps the id deterministic before the file is opened.
                    file_stem(&path),
                    path,
                    mtime,
                    size,
                ));
            }
        }
        out
    }

    fn read_into(&self, session: &SessionRef, since_offset: i64, sink: &mut dyn FnMut(Record)) {
        for (line_offset, raw_line) in JsonlLines::open(&session.file_path, since_offset) {
            // `since_offset == 0` means "fresh read, yield everything";
            // otherwise the caller already saw the record at exactly that
            // offset.
            if since_offset > 0 && line_offset <= since_offset {
                continue;
            }
            let stripped = py_bytes_strip(&raw_line);
            if stripped.is_empty() {
                continue;
            }
            // LOG: python debug-logs "Skipping malformed Qwen JSON line in %s".
            let Some(entry) = parse_json(stripped) else {
                continue;
            };
            // Valid JSON that is not an object cannot be a chat entry.
            if !entry.is_object() {
                continue;
            }
            if let Some(record) = record_from_entry(&entry, session, line_offset) {
                sink(record);
            }
        }
    }

    /// `backup create` copies the projects root (`source_roots`).
    ///
    /// Qwen declares `source_roots` and *not* `watch_paths`, so the ETL watcher
    /// gets nothing and falls back to periodic ingest while backup still copies
    /// the tree. Porting that asymmetry rather than "fixing" it keeps the
    /// watcher's provider set identical across the two implementations.
    fn source_roots(&self) -> Vec<PathBuf> {
        vec![self.root.clone()]
    }
}

/// One JSONL entry → a `Record`, or `None` (`_record_from_entry`).
fn record_from_entry(entry: &Value, session: &SessionRef, seq: i64) -> Option<Record> {
    let map = entry.as_object()?;
    // Only the two conversational types produce records; a non-string `type`
    // matches neither.
    let role = match map.get("type").and_then(Value::as_str) {
        Some(role @ ("user" | "assistant")) => role,
        _ => return None,
    };

    let parts = map
        .get("message")
        .filter(|value| value.is_object())
        .and_then(|message| message.get("parts"));
    let tokens = normalize_usage(map.get("usageMetadata"));
    let session_id = map
        .get("sessionId")
        .filter(|value| pyval::py_truthy(value))
        .map_or_else(|| session.session_id.clone(), pyval::py_str);

    Some(Record {
        provider: NAME.to_string(),
        session_id: session_id.clone(),
        seq,
        timestamp: map
            .get("timestamp")
            .filter(|value| pyval::py_truthy(value))
            .map_or_else(String::new, pyval::py_str),
        role: role.to_string(),
        // A non-string model would poison the Record contract downstream, so it
        // is treated as absent; an assistant turn then falls back to the
        // synthetic default and a user turn stays model-less.
        model: match map.get("model").and_then(Value::as_str) {
            Some(model) if !model.is_empty() => Some(model.to_string()),
            _ if role == "assistant" => Some(DEFAULT_MODEL.to_string()),
            _ => None,
        },
        input_tokens: tokens.input,
        output_tokens: tokens.output,
        cache_create_tokens: tokens.cache_creation,
        cache_read_tokens: tokens.cache_read,
        content_text: text_from_parts(parts),
        tools: tools_from_parts(parts),
        cwd: None,
        is_sidechain: false,
        uuid: map
            .get("uuid")
            .filter(|value| pyval::py_truthy(value))
            .map_or_else(|| format!("{session_id}:{seq}"), pyval::py_str),
        parent_uuid: None,
        raw: entry.clone(),
        speed: Speed::Standard,
    })
}

/// Concatenate every `.text` across content parts (`_text_from_parts`).
///
/// Thought parts keep their text: reasoning traces stay searchable, the same
/// call the Claude adapter makes for `thinking` blocks.
#[must_use]
pub fn text_from_parts(parts: Option<&Value>) -> String {
    let Some(blocks) = parts.and_then(Value::as_array) else {
        return String::new();
    };
    let mut pieces: Vec<String> = Vec::new();
    for part in blocks {
        if let Some(map) = part.as_object() {
            // A *missing*, empty, or non-string `text` contributes nothing at
            // all here — note the asymmetry with the Gemini adapter, where a
            // present-but-empty `text` still costs a newline in the join.
            if let Some(text) = map.get("text").and_then(Value::as_str)
                && !text.is_empty()
            {
                pieces.push(text.to_string());
            }
        } else if let Some(text) = part.as_str() {
            pieces.push(text.to_string());
        }
    }
    pieces.join("\n")
}

/// Tool names from `functionCall` blocks (`_tools_from_parts`).
#[must_use]
pub fn tools_from_parts(parts: Option<&Value>) -> Vec<String> {
    let Some(blocks) = parts.and_then(Value::as_array) else {
        return Vec::new();
    };
    let mut names = Vec::new();
    for part in blocks {
        let Some(name) = part
            .as_object()
            .and_then(|part| part.get("functionCall"))
            .filter(|value| value.is_object())
            .and_then(|call| call.get("name"))
            .and_then(Value::as_str)
            .filter(|name| !name.is_empty())
        else {
            continue;
        };
        names.push(
            TOOL_NAME_MAP
                .iter()
                .find(|(from, _)| *from == name)
                .map_or_else(|| name.to_string(), |(_, to)| (*to).to_string()),
        );
    }
    names
}

/// Flatten Qwen's `usageMetadata` into the canonical four slots
/// (`_normalize_usage`).
#[must_use]
pub fn normalize_usage(usage: Option<&Value>) -> crate::codex::CanonicalTokens {
    let Some(usage) = usage.and_then(Value::as_object) else {
        return crate::codex::CanonicalTokens {
            input: 0,
            output: 0,
            cache_creation: 0,
            cache_read: 0,
        };
    };
    canonicalize(usage)
}

fn canonicalize(usage: &Map<String, Value>) -> crate::codex::CanonicalTokens {
    let prompt = pyval::safe_int(usage.get("promptTokenCount"));
    let candidates = pyval::safe_int(usage.get("candidatesTokenCount"));
    let thoughts = pyval::safe_int(usage.get("thoughtsTokenCount"));
    let cached = pyval::safe_int(usage.get("cachedContentTokenCount"));
    crate::codex::CanonicalTokens {
        input: (prompt - cached).max(0),
        output: candidates.saturating_add(thoughts),
        cache_creation: 0,
        cache_read: cached,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn data_dir_env_relocates_the_projects_root() {
        let home = Path::new("/home/me");
        assert_eq!(
            resolve_projects_root(None, Some(home)),
            Path::new("/home/me/.qwen/projects")
        );
        assert_eq!(
            resolve_projects_root(Some(OsStr::new("/data/qwen")), Some(home)),
            Path::new("/data/qwen/projects")
        );
        assert_eq!(
            resolve_projects_root(Some(OsStr::new("")), Some(home)),
            Path::new("/home/me/.qwen/projects"),
            "an empty QWEN_DATA_DIR is Python-falsy"
        );
    }

    #[test]
    fn usage_flattens_cached_out_of_input_and_thoughts_into_output() {
        // The checked-in beta-normalizer fixture's first assistant turn.
        let tokens = normalize_usage(Some(&json!({
            "promptTokenCount": 1500,
            "candidatesTokenCount": 400,
            "thoughtsTokenCount": 100,
            "cachedContentTokenCount": 800,
        })));
        assert_eq!(tokens.input, 700);
        assert_eq!(tokens.output, 500);
        assert_eq!(tokens.cache_creation, 0);
        assert_eq!(tokens.cache_read, 800);
    }

    #[test]
    fn garbage_usage_degrades_to_zero_and_never_goes_negative() {
        let tokens = normalize_usage(Some(&json!({
            "promptTokenCount": "garbage",
            "cachedContentTokenCount": 99,
            "candidatesTokenCount": [1],
        })));
        assert_eq!(tokens.input, 0, "cached must never push input negative");
        assert_eq!(tokens.cache_read, 99);
        assert_eq!(tokens.output, 0);

        let missing = normalize_usage(None);
        assert_eq!(missing.input, 0);
        assert_eq!(missing.output, 0);
        let not_an_object = normalize_usage(Some(&json!("nope")));
        assert_eq!(not_an_object.cache_read, 0);
    }

    #[test]
    fn text_keeps_thought_parts_and_skips_empty_ones() {
        let parts = json!([
            {"text": "visible"},
            {"text": "", "thought": true},
            {"text": "reasoning", "thought": true},
            {"functionCall": {"name": "shell"}},
            "bare",
            42,
        ]);
        assert_eq!(text_from_parts(Some(&parts)), "visible\nreasoning\nbare");
        assert_eq!(text_from_parts(None), "");
        assert_eq!(text_from_parts(Some(&json!("not a list"))), "");
    }

    #[test]
    fn tool_names_map_and_unknown_ones_pass_through() {
        let parts = json!([
            {"functionCall": {"name": "shell"}},
            {"functionCall": {"name": "my_new_tool"}},
            {"functionCall": {"name": ""}},
            {"functionCall": "not an object"},
            {"functionCall": {"name": 7}},
        ]);
        assert_eq!(
            tools_from_parts(Some(&parts)),
            vec!["Bash".to_string(), "my_new_tool".to_string()]
        );
        assert!(tools_from_parts(None).is_empty());
    }
}
